File: Assets/Scripts/NeuralNetwork/NeuralNetwork.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using Random = UnityEngine.Random;

namespace Assets.Scripts.NeuralNetwork
{
    [Serializable]
    public class NeuralNetwork
    {
        public double Fitness;
        public NN Network;

        public NeuralNetwork(double learningRate, int inputCount, int hiddenLayerCount, int outputCount)
        {
            Fitness = 0;
            Network.LearningRate = learningRate;
            BuildInitialNeuralNetwork(new[] {inputCount, hiddenLayerCount, outputCount});
        }

        private void BuildInitialNeuralNetwork(IList<int> layers)
        {
            Network.Layers = new Layer[layers.Count];

            for (var layer = 0; layer < layers.Count; ++layer)
            {
                var count = layers[layer];
                Network.Layers[layer].Neurons = new Neuron[count];

                for (var neuron = 0; neuron < count; ++neuron)
                {
                    if (layer == 0) continue;

                    Network.Layers[layer].Neurons[neuron].Bias = NextRandom();
                    Network.Layers[layer].Neurons[neuron].Dendrite = new Dendrite[layers[layer - 1]];
                    for (var dendrite = 0; dendrite < Network.Layers[layer].Neurons[neuron].Dendrite.Length; dendrite++)
                    {
                        Network.Layers[layer].Neurons[neuron].Dendrite[dendrite].Weight = NextRandom();
                    }
                }
            }
        }

        private static double NextRandom()
        {
            return Random.Range(1f, 2f);
        }

        public double[] GetWeights()
        {
            return new double[2];
        }

        public double BipolarSigmoid(double x)
        {
            return (1/(1 + Math.Exp(x*-1)));
        }

        public void SetWeights(double[] childWeights)
        {
        }

        public int DendritesCount()
        {
            return 0;
        }

        public NeuralOutput Think(SensoryInput input)
        {
            for (var layer = 0; layer < Network.Layers.Length; ++layer)
            {
                for (var neuron = 0; neuron < Network.Layers[layer].Neurons.Length; neuron++)
                {
                    if (layer == 0)
                        Network.Layers[layer].Neurons[neuron].Value = input.Values[neuron];
                    else
                    {
                        Network.Layers[layer].Neurons[neuron].Value = 0;
                        for (var dentrite = 0; dentrite < Network.Layers[layer - 1].Neurons.Length; ++dentrite)
                        {
                            Network.Layers[layer].Neurons[neuron].Value = Network.Layers[layer].Neurons[neuron].Value +
                                                                          Network.Layers[layer - 1].Neurons[dentrite]
                                                                              .Value*
                                                                          Network.Layers[layer].Neurons[neuron].Dendrite
                                                                              [dentrite].Weight;
                        }
                        Network.Layers[layer].Neurons[neuron].Value =
                            BipolarSigmoid(Network.Layers[layer].Neurons[neuron].Value);
                    }
                }
            }

            NeuralOutput output;
            output.Values = new double[Network.Layers[Network.Layers.Length - 1].Neurons.Length];
            for (var neuron = 0; neuron < output.Values.Length; ++neuron)
            {
                output.Values[neuron] =
                    Mathf.Clamp((float) Network.Layers[Network.Layers.Length - 1].Neurons[neuron].Value, -1, 1);
            }

            return output;
        }
    }
}

File: js/NeuralNetwork.js
import { hyperbolicTangent, randomRange } from './utils.js';

export class NeuralNetwork {
  constructor(inputCount, hiddenLayerCount, outputCount) {
    this.layers = [];
    const layerSizes = [inputCount, hiddenLayerCount, outputCount];

    for (let i = 0; i < layerSizes.length; i++) {
      const layer = { neurons: [] };

      for (let j = 0; j < layerSizes[i]; j++) {
        const neuron = { value: 0, bias: 0, dendrites: [] };

        if (i > 0) {
          const fanIn = layerSizes[i - 1];
          const fanOut = layerSizes[i];
          const limit = Math.sqrt(6 / (fanIn + fanOut));
          neuron.bias = randomRange(-limit, limit);

          for (let k = 0; k < fanIn; k++) {
            neuron.dendrites.push({ weight: randomRange(-limit, limit) });
          }
        }

        layer.neurons.push(neuron);
      }

      this.layers.push(layer);
    }
  }

  think(inputValues) {
    for (let i = 0; i < this.layers[0].neurons.length; i++) {
      this.layers[0].neurons[i].value = inputValues[i];
    }

    for (let layer = 1; layer < this.layers.length; layer++) {
      for (let neuron = 0; neuron < this.layers[layer].neurons.length; neuron++) {
        const n = this.layers[layer].neurons[neuron];
        let sum = n.bias;

        for (let p = 0; p < this.layers[layer - 1].neurons.length; p++) {
          sum += this.layers[layer - 1].neurons[p].value * n.dendrites[p].weight;
        }

        n.value = hyperbolicTangent(sum);
      }
    }

    return this.layers[this.layers.length - 1].neurons.map(n => n.value);
  }

  getWeights() {
    const weights = [];
    for (let layer = 1; layer < this.layers.length; layer++) {
      for (let neuron = 0; neuron < this.layers[layer].neurons.length; neuron++) {
        const n = this.layers[layer].neurons[neuron];
        weights.push(n.bias);
        for (let d = 0; d < n.dendrites.length; d++) {
          weights.push(n.dendrites[d].weight);
        }
      }
    }
    return weights;
  }

  setWeights(weights) {
    let idx = 0;
    for (let layer = 1; layer < this.layers.length; layer++) {
      for (let neuron = 0; neuron < this.layers[layer].neurons.length; neuron++) {
        const n = this.layers[layer].neurons[neuron];
        n.bias = weights[idx++];
        for (let d = 0; d < n.dendrites.length; d++) {
          n.dendrites[d].weight = weights[idx++];
        }
      }
    }
  }

  getDendriteCount() {
    let count = 0;
    for (let layer = 1; layer < this.layers.length; layer++) {
      for (let neuron = 0; neuron < this.layers[layer].neurons.length; neuron++) {
        count += this.layers[layer].neurons[neuron].dendrites.length;
      }
    }
    return count;
  }

  clone() {
    const inputCount = this.layers[0].neurons.length;
    const hiddenCount = this.layers[1].neurons.length;
    const outputCount = this.layers[2].neurons.length;
    const copy = new NeuralNetwork(inputCount, hiddenCount, outputCount);
    copy.setWeights(this.getWeights());
    return copy;
  }
}
